Add comparison of INT8 and FP32 models

nikita-savelyevv opened the pull request:
Added the following features to PyTorch and Tensorflow quantization aware training notebooks
- fine-tuning of float32 model the same way as int8 model is finetuned
- accuracy comparison between fine-tuned int8 and fine-tuned float32 models

Note: nbval fails, however it also seems to fail for the `main` branch

negvet commented:
Not sure if I understand. Please explain what is the reason for the fp32 model finetuning. 
It makes sense if we train the full precision model from scratch in the notebook. But instead, we provide the weights of the fp32 model anyway, so why not finetune it beforehand in the same way as the int8 model is finetuned?

Asking because it makes the tutorial longer and probably harder to read through.

nikita-savelyevv commented:
Thanks!

I started with an assumption that it would be beneficial for a user of the notebook to see how INT8 model accuracy is compared to FP32 model accuracy. As of now, we can only justifiably compare their accuracies right after initialization of INT8 model. 

The initialized INT8 model then gets fine-tuned to restore the accuracy drop. However, it would be wrong to compare fine-tuned INT8 model to the original FP32 model, because the former one was effectively trained on a bigger dataset at this point. By fine-tuning the FP32 model the same way, it is now correct to compare the fine-tuned INT8 model with it. Because they are both trained on the same amount of data.

When executing, this experiment shows that both fine-tuned models reach comparable accuracies, and hence we demonstrate that fine-tuning FP32 model on the same amount of data does not give marginally superior accuracy over the fine-tuned INT8 model (I guess it actually could).

But I have to agree that this comes with a price of increased number of source lines and time to execute.

ljaljushkin commented:
@nikita-savelyevv could you please provide an example of typical accuracies of tuned and original models with your changes?
int8 model should be still within 1% accuracy drop of tuned fp32 one. 
and tuned fp32 shouldn't degrade a lot in comparison with original fp32.

nikita-savelyevv commented:
Performed four experiments for each PyTorch and Tensorflow. The averaged results are the following.

                                           Torch
INT8 base     FP32 base     INT8 tuned    FP32 tuned
     53.34            53.24                54.72             54.87

                                               TF
INT8 base      FP32 base    INT8 tuned    FP32 tuned
     78.45           78.2                80.05             80.35

In most cases initialized INT8 is better compared to the loaded FP32 and fine-tuned FP32 is better than fine-tuned INT8, but not always.

alexsu52 reviewed (changes requested):
Please update pre-trained weights not a code of the notebook.

helena-intel commented:
> The initialized INT8 model then gets fine-tuned to restore the accuracy drop. However, it would be wrong to compare fine-tuned INT8 model to the original FP32 model, because the former one was effectively trained on a bigger dataset at this point. By fine-tuning the FP32 model the same way, it is now correct to compare the fine-tuned INT8 model with it. Because they are both trained on the same amount of data.

That may be true from an academic perspective, but is this relevant for users? Users train a model until desired accuracy, want to quantize it and since keeping accuracy is important to them, they choose to do that with NNCF by finetuning the model. It is important to them that the quantized model does not have a significantly worse accuracy, and that's the reason the accuracy comparison is shown in the notebook.

helena-intel reviewed (approved):
Thank you @nikita-savelyevv this is much appreciated, and I'm happy with the improvements and clarifications. I have not run the notebooks but the changes look good to me.

@paularamo Assuming no review issues, I would recommend to merge this PR before the 2022.1 release, if possible. The only change required for the 2.0 API should be the last line, to check for the name of the CPU device (and that's not an essential change - the existing line will work in 2022.1 too).

paularamo commented:
I recommend updating the branch. We could merge, but we will need your help during the migration if that is needed.

alexsu52 commented on notebooks/302-pytorch-quantization-aware-training/302-pytorch-quantization-aware-training.ipynb:
```diff
@@ -697,9 +723,10 @@
     "\n",
-    "print(f\"Accuracy of tuned INT8 model: {acc1:.3f}\")"
+    "print(f\"Accuracy of tuned INT8 model: {acc1_int8:.3f}\")\n",
+    "print(f\"Accuracy gain/drop of tuned INT8 model over pre-trained FP32 model: {acc1_int8 - acc1_fp32:.3f}\")"
```
```suggestion
    "print(f\"Accuracy drop of tuned INT8 model over pre-trained FP32 model: {acc1_fp32 - acc1_int8:.3f}\")"
```

alexsu52 commented on notebooks/302-pytorch-quantization-aware-training/302-pytorch-quantization-aware-training.ipynb:
```diff
@@ -872,4 +899,4 @@
  "nbformat": 4,
  "nbformat_minor": 4
-}
+}
```
Please, revert changes.

alexsu52 commented on notebooks/305-tensorflow-quantization-aware-training/305-tensorflow-quantization-aware-training.ipynb:
```diff
@@ -557,4 +558,4 @@
  "nbformat": 4,
  "nbformat_minor": 5
-}
+}
```
Please, revert changes.

alexsu52 commented on notebooks/305-tensorflow-quantization-aware-training/305-tensorflow-quantization-aware-training.ipynb:
```diff
@@ -402,9 +402,10 @@
     "                                     callbacks=tf.keras.callbacks.ProgbarLogger(stateful_metrics=['acc@1']))\n",
-    "print(f\"\\nAccuracy of INT8 model after fine-tuning: {test_acc:.3f}\")"
+    "print(f\"\\nAccuracy of INT8 model after fine-tuning: {test_acc:.3f}\")\n",
+    "print(f\"\\nAccuracy gain/drop of tuned INT8 model over pre-trained FP32 model: {acc_int8 - acc_fp32:.3f}\")"
```
```suggestion
    "print(f\"\\nAccuracy drop of tuned INT8 model over pre-trained FP32 model: {acc_fp32 - acc_int8:.3f}\")"
```

ryanloney commented:
@ljaljushkin you are listed as approver. Is this ok to merge?

nikita-savelyevv commented:
@ryanloney, ready to merge, thanks!

ryanloney commented:
@Debskij can you see why the CI check failed on one of the tests with ubuntu 18? Might need to re-run. 

Let's prepare to merge this on Wednesday.

nikita-savelyevv commented:
Pushed some additional changes. Thus, the total list of changes for this PR is the following:

- added explicit comparison of baseline pre-trained FP32 model with fine-tuned INT8 model
- tweaked the comment above the INT8 model fine-tuning to better fit what really happens there
- fixed a newly found bug with model validation for notebook 302 (image augmentation was active during validation), switched to Tiny Imagenet own validation set instead of splitting it from the train set
- updated the models for both PyTorch and Tensorflow with the more accurate ones
- added random seed setting for reproducibility purposes

nikita-savelyevv commented:
> @Debskij can you see why the CI check failed on one of the tests with ubuntu 18? Might need to re-run.
> 
> Let's prepare to merge this on Wednesday.

I've added some additional commits since, but now from our side the code is ready for merge

Debskij commented:
Hi @nikita-savelyevv! 
Thanks for this PR, please rebase your branch to apply recent commits

nikita-savelyevv commented:
> Hi @nikita-savelyevv! Thanks for this PR, please rebase your branch to apply recent commits

Done!

nikita-savelyevv commented:
@Debskij Rebased once again, should be ready for merge
